stdlib/public/core/StringBridge.swift

```diff
@@ -17,34 +17,113 @@ import SwiftShims
 internal typealias _CocoaString = AnyObject
 
 #if _runtime(_ObjC)
+
 // Swift's String bridges NSString via this protocol and these
 // variables, allowing the core stdlib to remain decoupled from
 // Foundation.
 
+@objc internal protocol _StringSelectorHolder : _NSCopying {
+    
+  @objc var length: Int { get }
+  
+  @objc var hash: UInt { get }
+  
+  @objc(characterAtIndex:)
+  func character(at offset: Int) -> UInt16
+  
+  @objc(getCharacters:range:)
+  func getCharacters(
+   _ buffer: UnsafeMutablePointer<UInt16>, range aRange: _SwiftNSRange
+  )
+ 
+  @objc(_fastCStringContents:)
+  func _fastCStringContents(
+    _ requiresNulTermination: Int8
+  ) -> UnsafePointer<CChar>?
+  
+  @objc(_fastCharacterContents)
+  func _fastCharacterContents() -> UnsafePointer<UInt16>?
+  
+  @objc(getBytes:maxLength:usedLength:encoding:options:range:remainingRange:)
+  func getBytes(_ buffer: UnsafeMutableRawPointer?,
+   maxLength maxBufferCount: Int,
+  usedLength usedBufferCount: UnsafeMutablePointer<Int>?,
+    encoding: UInt,
+     options: UInt,
+       range: _SwiftNSRange,
+       remaining leftover: UnsafeMutablePointer<_SwiftNSRange>?) -> Int8
+  
+  @objc(compare:options:range:locale:)
+  func compare(_ string: _CocoaString,
+               options: UInt,
+               range: _SwiftNSRange,
+               locale: AnyObject?) -> Int
+}
+
+/*
+ Passing a _CocoaString through _objc() lets you call ObjC methods that the
+ compiler doesn't know about, via the protocol above. In order to get good
+ performance, you need a double indirection like this:
+ 
+  func a -> _objc -> func a'
+ 
+ because any refcounting @_effects on 'a' will be lost when _objc breaks ARC's
+ knowledge that the _CocoaString and _StringSelectorHolder are the same object.
+ */
+@inline(__always)
+internal func _objc(_ str: _CocoaString) -> _StringSelectorHolder {
```

Comment:
Does declaring this as a freestanding function benefit us in any way? At least for organization, it'd be nice to have this as an extension on `_CocoaString`.

Reply:
also discussed in person